File: nCr.java
import java.io.*;
import java.util.*;
class nCr {
	static int t,n,cnt_o,cnt_z,sum;
	static int maxn = 1000010;
	static long inp[] = new long[maxn];
	static long mod = 1000000007;
	static long mul[] = new long[maxn];
	static long power (long a, long b) {
		long res = 1;
		while (b > 0) {
			if(b % 2 == 1)
				res = (res * a + mod) % mod;
			b = b >> 1;
			a = (a * 1L * a + mod) % mod;
		}
		return res % mod;
	}
	static void pre () {
		mul[1] = 1;
		for (int i = 2; i < maxn; ++i)
			mul[i] = (mul[i - 1] * i % mod + mod) % mod;
	}
	static long nCr(int n_, int r) {
		if (n_ - r >= 0)
	   		r = Math.min (r, n_ - r);
	   	if (r == 0)
	      return 1;
	    long a = mul[n_];
	    long b = mul[r];
	    long c = mul[n_ - r];
	    long d = power(b, mod - 2) % mod;
	    long e = power(c, mod - 2) % mod;
	    //System.out.println (a + " " + b + " " + c + " " + d + " " + e);
	    return (mul[n_] % mod * 1L * d % mod * 1L * e % mod) % mod;
	}
	public static void main(String args[]) {
		pre();
		Scanner ss = new Scanner (System.in);
		BufferedReader br = new BufferedReader (new InputStreamReader (System.in));
		t = ss.nextInt ();
		while(t>0) {
			t--;
			int a,b;
			a = ss.nextInt ();
			b = ss.nextInt ();
			System.out.println (nCr(a, b) % mod);
		}
	}
}

File: nCr.cpp
#include <bits/stdc++.h>
using namespace std;
#define mod 1000000007
#define maxn 1000010
#define ll long long int
ll n, t, M;
ll num[maxn], rem[maxn];
ll mul_inv(ll a, ll m) {
	ll t, y = 0, x = 1, m0 = m, q;
	while (a > 1) {
		q = a / m;
		t = m;
		m = a % m;
		a = t;
		t = y;
		y = x - q * y;
		x = t;
	}
	if (x < 0)
		x += m0;
	return x;
}
ll power(ll x, ll y) {
	ll res = 1;
	while (y) {
		if (y & 1) res = res * x;
		x = x * x;
		y >>= 1;
	}
	return res;
}
ll chinese() {
	ll prod = 1;
	ll sum = 0;
	for (int i = 0; i < n; ++i)
		prod *= num[i];
	for (int i = 0; i < n; ++i) {
		M = prod / num[i];
		sum = sum + rem[i] * M * mul_inv(M, num[i]);
	}
	return sum % prod;
}
int main() {
	scanf("%lld", &t);
	while (t--) {
		scanf("%lld", &n);
		for (int i = 0; i < n; ++i)
			scanf("%lld", &num[i]);
		for (int i = 0; i < n; ++i)
			scanf("%lld", &rem[i]);
		printf("%lld\n", chinese());
	}
	return 0;
}